### DISCUSSION TITLE
[metadata.ownerReferences] Error to create the deployment and service with custom CRDs

### DISCUSSION THREAD

**Discussion by Tomelin** [kubernetes-sigs/kubebuilder#3160] [Support] [answered]
Hi guys,

I'm trying to create a CRDs but I'm having errors when creating the deployment and service.

**the error is:**
 _Error to create the service, Service "sc-monogodb" is invalid: [[metadata.ownerReferences.name](http://metadata.ownerreferences.name/): Invalid value: "": name must not be empty, metadata.ownerReferences.uid: Invalid value: "": uid must not be empty]1.673797609427857e+09 ERROR controller.discovery Reconciler error {"reconciler group": "catalog.mydomain.io", "reconciler kind": "Discovery", "name": "discovery-sample", "namespace": "service-discovery", "error": "Service \"sc-monogodb\" is invalid: [[metadata.ownerReferences.name](http://metadata.ownerreferences.name/): Invalid value: \"\": name must not be empty, metadata.ownerReferences.uid: Invalid value: \"\": uid must not be empty]"}_

My parameters are:
```
	d *catalogv1alpha1.Discovery
	svc := &corev1.Service{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "sc-monogodb",
			Namespace: "service-discovery",
			OwnerReferences: []metav1.OwnerReference{{
				Name:       "discovery-sample",
				UID:        "b4ebb054-2e65-4016-bc04-cc3aad91d644",
				APIVersion: "catalog.mydomain.io/v1alpha1",
				Kind:       "Discovery",
			}},
		},
		Spec: corev1.ServiceSpec{
			Ports: []corev1.ServicePort{{
				Name:     "sc-mongodb",
				Protocol: corev1.ProtocolTCP,
				Port:     27017,
			}},
		},
	}
```

I changed the parameters OwnerReferences to variable and not worked.

I'm using this kubebuilder version:
 kubebuilder version
Version: main.version{KubeBuilderVersion:"0.0.0-SNAPSHOT-fc59b56", KubernetesVendor:"1.23.5", GitCommit:"fc59b569ced3a2942b87ce20db39043be1bf8352", BuildDate:"2022-05-18T10:55:28Z", GoOs:"darwin", GoArch:"amd64"}

How can I set the Owner reference in the deployment/service?

Thanks!

**Comment by Tomelin** [comment]
Hi, 

I managed to adjust the code.  The problem was I didn't call the function "r.Get(ctx, req, d)"  to read the information from my Kind.

I adjust it and worked.

Thank you!

**Comment by camilamacedo86** [comment, marked as answer]
Hi @Tomelin ,

The error you're encountering seems came from the fact that the `name` and `uid` fields in the `OwnerReferences` shouldn't be empty. Using the `controller-runtime` library, which comes bundled with `kubebuilder`, there's a more straightforward way to set the owner reference correctly.

Here's how you can do it:

- Fetch the actual `Discovery` custom resource (CR) instance that you want to set as the owner of the `Service` object.

-  Use the `controllerutil.SetOwnerReference` method to correctly set the owner reference.

```go
import (
	...
	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
)

// Assuming you've fetched the actual Discovery CR into a variable named 'd'
svc := &corev1.Service{
	ObjectMeta: metav1.ObjectMeta{
		Name:      "sc-monogodb",
		Namespace: "service-discovery",
	},
	Spec: corev1.ServiceSpec{
		Ports: []corev1.ServicePort{{
			Name:     "sc-mongodb",
			Protocol: corev1.ProtocolTCP,
			Port:     27017,
		}},
	},
}

// Set the owner reference
if err := controllerutil.SetControllerReference(d, svc, mgr.GetScheme()); err != nil {
	return err
}
```

I hope that answer your question.
Please, ensure that you:

- check the scaffold done by the plugin Deploy Image: https://book.kubebuilder.io/plugins/deploy-image-plugin-v1-alpha
- check how the controller is implemented to see a good example: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/internal/controller/busybox_controller.go
- Also, check the content of the PR https://github.com/kubernetes-sigs/kubebuilder/pull/3596/files. I think it can will help you out either understand how things work better. 

I am closing this one as sorted out
However, feel free to reopen if you see that fits.
